### PR TITLE
elide partial bounds check in list iterator

### PR DESCRIPTION
**Pull request by ritchie46**
This PR elides some of the bounds checks done while iteration list values. The slicing still does bound checks and IMO should be one of the invariants of the offsets (being in bounds). I want to propose a `slice_unchecked` on the `Array` trait. Let me know what you think.

This optimization might also work for `FixedSizeList` if we add a `value_unchecked` to the array.

```
Gnuplot not found, using plotters backend
list: iter_values 2^10  time:   [15.214 us 15.217 us 15.220 us]                                    
                        change: [-12.125% -12.063% -11.995%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  2 (2.00%) high mild
  8 (8.00%) high severe

list: iter 2^10         time:   [15.274 us 15.277 us 15.280 us]                             
                        change: [-12.382% -12.065% -11.801%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

list: iter_values 2^12  time:   [60.861 us 60.869 us 60.878 us]                                   
                        change: [-11.760% -9.6523% -7.8631%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  7 (7.00%) high mild
  3 (3.00%) high severe

list: iter 2^12         time:   [61.928 us 61.995 us 62.076 us]                            
                        change: [-5.0382% -4.4427% -3.8211%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

list: iter_values 2^14  time:   [244.38 us 244.41 us 244.44 us]                                   
                        change: [-8.0274% -7.9989% -7.9730%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  1 (1.00%) high mild
  5 (5.00%) high severe

list: iter 2^14         time:   [245.20 us 245.23 us 245.26 us]                            
                        change: [-21.663% -18.963% -16.142%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

Benchmarking list: iter_values 2^16: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.4s, enable flat sampling, or reduce sample count to 60.
list: iter_values 2^16  time:   [1.0023 ms 1.0034 ms 1.0051 ms]                                    
                        change: [-3.7442% -3.3126% -2.7332%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe

Benchmarking list: iter 2^16: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.2s, enable flat sampling, or reduce sample count to 60.
list: iter 2^16         time:   [1.0243 ms 1.0254 ms 1.0266 ms]                             
                        change: [-0.3685% -0.1351% +0.1820%] (p = 0.35 > 0.05)
                        No change in performance detected.
Found 19 outliers among 100 measurements (19.00%)
  8 (8.00%) low mild
  8 (8.00%) high mild
  3 (3.00%) high severe

list: iter_values 2^18  time:   [3.9003 ms 3.9005 ms 3.9008 ms]                                    
                        change: [-11.058% -10.852% -10.638%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

list: iter 2^18         time:   [3.9241 ms 3.9244 ms 3.9248 ms]                             
                        change: [-6.2696% -6.1584% -6.0561%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

list: iter_values 2^20  time:   [15.604 ms 15.606 ms 15.607 ms]                                   
                        change: [-7.4029% -7.2619% -7.1323%] (p = 0.00 < 0.05)
                        Performance has improved.

list: iter 2^20         time:   [15.923 ms 16.046 ms 16.204 ms]                            
                        change: [-4.1817% -3.4015% -2.4924%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
```

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
Thanks!

Yeap, unchecked slicing makes sense 👍

File: benches/iter_list.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        let size = 2usize.pow(log2_size);
+
+        let values = Buffer::from_iter(0..size as i32);
+        let values = PrimitiveArray::<i32>::from_data(DataType::Int32, values, None);
+
+        let mut offsets = MutableBuffer::from_iter((0..size as i32).step_by(2));
+        offsets.push(size as i32);
+
+        let data_type = ListArray::<i32>::default_datatype(DataType::Int32);
+        let array = ListArray::<i32>::from_data(data_type, offsets.into(), Arc::new(values), None);
```
Comment (jorgecarleitao) [inline benches/iter_list.rs:23]: should we add a validity here so that the bench `iter` hits null slots?

Comment (ritchie46) [reply benches/iter_list.rs:23]: :+1:

**Comment by ritchie46** [comment]
@jorgecarleitao  I also added the `unchecked_slice` here. This reduces the bounds checking in safe code from two checks to a single one. 

Now the bounds are asserted `array::slice` and the implementation is dispatched to the `unchecked` function.
Before the bounds were asserted in `buffer::slice` and in `validity::slice`.

File: src/buffer/immutable.rs [hunk 2]
```diff
@@ -86,11 +86,22 @@ impl<T: NativeType> Buffer<T> {
...
     /// Panics iff `offset` is larger than `len`.
     #[inline]
-    pub fn slice(mut self, offset: usize, length: usize) -> Self {
+    pub fn slice(self, offset: usize, length: usize) -> Self {
         assert!(
             offset + length <= self.len(),
             "the offset of the new Buffer cannot exceed the existing length"
         );
+        // Safety:
+        // we just check bounds
```
Comment (jorgecarleitao) [inline src/buffer/immutable.rs:95]:
```suggestion
        // Safety: we just checked bounds
```

File: src/array/utf8/mod.rs [hunk 3]
```diff
@@ -165,9 +165,24 @@ impl<O: Offset> Utf8Array<O> {
...
+            offset + length <= self.len(),
+            "the offset of the new Buffer cannot exceed the existing length"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+    /// Returns a slice of this [`Utf8Array`].
+    /// # Implementation
+    /// This operation is `O(1)` as it amounts to essentially increase two ref counts.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/utf8/mod.rs:178]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/union/mod.rs [hunk 4]
```diff
@@ -190,6 +190,23 @@ impl UnionArray {
             offset: self.offset + offset,
         }
     }
+
+    /// Returns a slice of this [`UnionArray`].
+    /// # Implementation
+    /// This operation is `O(F)` where `F` is the number of fields.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/union/mod.rs:198]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/struct_.rs [hunk 5]
```diff
@@ -105,13 +105,29 @@ impl StructArray {
...
+            "offset + length may not exceed length of array"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    /// Creates a new [`StructArray`] that is a slice of `self`.
+    /// # Implementation
+    /// This operation is `O(F)` where `F` is the number of fields.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/struct_.rs:119]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/primitive/mod.rs [hunk 6]
```diff
@@ -86,10 +86,27 @@ impl<T: NativeType> PrimitiveArray<T> {
...
+            "offset + length may not exceed length of array"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    /// Returns a slice of this [`PrimitiveArray`].
+    /// # Implementation
+    /// This operation is `O(1)` as it amounts to increase two ref counts.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/primitive/mod.rs:100]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/mod.rs [hunk 7]
```diff
@@ -426,7 +434,9 @@ fn display_fmt<T: std::fmt::Display, I: IntoIterator<Item = Option<T>>>(
 
 /// Trait that list arrays implement for the purposes of DRY.
 pub trait IterableListArray: Array {
-    fn value(&self, i: usize) -> Box<dyn Array>;
+    // # Safety:
+    // The caller must ensure that `i < self.len()`
```
Comment (jorgecarleitao) [inline src/array/mod.rs:438]:
```suggestion
    /// # Safety:
    /// The caller must ensure that `i < self.len()`
```

File: src/array/mod.rs [hunk 8]
```diff
@@ -88,6 +88,14 @@ pub trait Array: std::fmt::Debug + Send + Sync {
     /// This function panics iff `offset + length >= self.len()`.
     fn slice(&self, offset: usize, length: usize) -> Box<dyn Array>;
 
+    /// Slices the [`Array`], returning a new `Box<dyn Array>`.
+    /// # Implementation
+    /// This operation is `O(1)` over `len`, as it amounts to increase two ref counts
+    /// and moving the struct to the heap.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`
```
Comment (jorgecarleitao) [inline src/array/mod.rs:96]:
```suggestion
    /// The caller must ensure that `offset + length <=| self.len()`
```

File: src/array/fixed_size_list/mod.rs [hunk 9]
```diff
@@ -63,12 +63,30 @@ impl FixedSizeListArray {
     /// Returns a slice of this [`FixedSizeListArray`].
     /// # Implementation
     /// This operation is `O(1)`.
+    /// # Panics
+    /// panics iff `offset + length >= self.len()`
```
Comment (jorgecarleitao) [inline src/array/fixed_size_list/mod.rs:67]:
```suggestion
    /// panics iff `offset + length > self.len()`
```

File: src/array/fixed_size_list/mod.rs [hunk 10]
```diff
@@ -63,12 +63,30 @@ impl FixedSizeListArray {
...
+            "the offset of the new Buffer cannot exceed the existing length"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    /// Returns a slice of this [`FixedSizeListArray`].
+    /// # Implementation
+    /// This operation is `O(1)`.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/fixed_size_list/mod.rs:80]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/fixed_size_binary/mod.rs [hunk 11]
```diff
@@ -50,12 +50,30 @@ impl FixedSizeBinaryArray {
...
+            "the offset of the new Buffer cannot exceed the existing length"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    /// Returns a slice of this [`FixedSizeBinaryArray`].
+    /// # Implementation
+    /// This operation is `O(1)` as it amounts to increase 3 ref counts.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/fixed_size_binary/mod.rs:67]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/fixed_size_binary/mod.rs [hunk 12]
```diff
@@ -50,12 +50,30 @@ impl FixedSizeBinaryArray {
     /// Returns a slice of this [`FixedSizeBinaryArray`].
     /// # Implementation
     /// This operation is `O(1)` as it amounts to increase 3 ref counts.
+    /// # Panics
+    /// panics iff `offset + length >= self.len()`
```
Comment (jorgecarleitao) [inline src/array/fixed_size_binary/mod.rs:54]:
```suggestion
    /// panics iff `offset + length > self.len()`
```

File: src/array/boolean/mod.rs [hunk 13]
```diff
@@ -61,10 +61,27 @@ impl BooleanArray {
...
+            "the offset of the new Buffer cannot exceed the existing length"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    /// Returns a slice of this [`BooleanArray`].
+    /// # Implementation
+    /// This operation is `O(1)` as it amounts to increase two ref counts.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/boolean/mod.rs:75]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

File: src/array/binary/mod.rs [hunk 14]
```diff
@@ -90,8 +90,24 @@ impl<O: Offset> BinaryArray<O> {
...
+            "the offset of the new Buffer cannot exceed the existing length"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    /// Creates a new [`BinaryArray`] by slicing this [`BinaryArray`].
+    /// # Implementation
+    /// This function is `O(1)`: all data will be shared between both arrays.
+    /// # Safety
+    /// The caller must ensure that `offset + length < self.len()`.
```
Comment (jorgecarleitao) [inline src/array/binary/mod.rs:104]:
```suggestion
    /// The caller must ensure that `offset + length <= self.len()`.
```

**Review by jorgecarleitao — COMMENTED** [review]
Went through it carefully and all checks great. Left some nits to apply, but it is good to go.

Thanks a lot, @ritchie46 ! 💯

**Pull request merged**